devito/arguments.py

```diff
@@ -158,198 +139,381 @@ class PtrArgument(Argument):
 
     is_PtrArgument = True
 
-    def __init__(self, name, provider):
-        super(PtrArgument, self).__init__(name, provider, provider.value)
-
-    def verify(self, value):
-        self._value = value or self._value
-        return True
-
-    def reset(self):
-        if callable(self.default_value):
-            self._value = self.default_value()
-        else:
-            self._value = self.default_value
+    def __init__(self, provider):
+        super(PtrArgument, self).__init__(provider.name,
+                                          [Dependency("gets_value_from", provider)],
+                                          provider.dtype)
 
 
-class ArgumentProvider(object):
-
-    """ Abstract base class for any object that, post code-generation, might resolve
-        resolve to runtime arguments. We assume that one source object (e.g. Dimension,
-        SymbolicData) might provide multiple runtime arguments.
+class ArgumentEngine(object):
+    """ Class that encapsulates the argument derivation and verification subsystem
     """
+    def __init__(self, stencils, parameters, dle_arguments):
+        self.stencils = stencils
+        self.parameters = parameters
+        self.dle_arguments = dle_arguments
+        self.argument_mapper = self._build_argument_mapper(parameters)
+        self.arguments = filter_ordered([x for x in self.argument_mapper
+                                         if isinstance(x, Argument)],
+                                        key=lambda x: x.name)
+        self.dims = [x for x in self.argument_mapper if isinstance(x, DimensionParameter)]
+        self.offsets = {d.end_name: v for d, v in retrieve_offsets(stencils).items()}
+
+    def handle(self, **kwargs):
+        """ The main method by which the :class:`Operator` interacts with this class.
+            The arguments passed into Operator.apply() all end up in kwargs here.
+        """
 
-    def __init__(self, *args, **kwargs):
-        pass
-
-    @property
-    def rtargs(self):
-        """ Return a list of RuntimeArgument objects that this object needs.
+        user_autotune = kwargs.pop('autotune', False)
+
+        kwargs = self._offset_adjust(kwargs)
+
+        kwargs = self._extract_children_of_composites(kwargs)
+
+        values = self._derive_values(kwargs)
+
+        # The following is only being done to update the autotune flag. The actual value
+        # derivation for the dle arguments has moved inside the above _derive_values
+        # method.
+        # TODO: Refactor so this is not required
+        dim_sizes = dict([(d.name, runtime_dim_extent(d, values))
+                          for d in self.dimensions])
+        dle_arguments, dle_autotune = self._dle_arguments(dim_sizes)
+
+        assert(self._verify(values))
+
+        arguments = OrderedDict([(k.name, v) for k, v in values.items()])
+
+        return arguments, user_autotune and dle_autotune
+
+    def _offset_adjust(self, kwargs):
```

Comment:
``return {k: v + self.offsets.get(k, 0) for k, v in kwargs.items()}``

Reply:
so this can probably be inlined

Reply:
Fixed

Reply:
not seeing it...

Reply:
Tried it. It won't work. 0 is not a good default value because not all arguments support the `+` operation.

Reply:
still, I encourage to make it pythonic:
``return {k: (v + self.offsets[k]) if k in self.offsets else v for k, v in kwargs.items()}``